### ISSUE TITLE
[bug] `tools.build:defines` is broken in multiconfig setting

### ISSUE THREAD

**Issue by stevenwdv** [conan-io/conan#15921]
### Describe the bug

`tools.build:defines` is broken in multiconfig setting, e.g. with MSVC. The quoting in the generator expression is not correct in `blocks.py`.

### How to reproduce it

- Set `tools.build:defines=['A=1', 'B=2']`
- Observe how the generated toolchain file or afterwards generated vxproj by CMake does not contain the right definitions.

EDIT: See PR #15924 for more info

**Comment by stevenwdv** [comment]
I'm working on a PR

**Comment by memsharded** [comment]
Thanks for the report and the offer to fix it @stevenwdv 

This was tested in ``test_cmake_toolchain_cxxflags_multi_config``, maybe it is because the single quotes vs double quotes?

**Comment by stevenwdv** [comment]
> maybe it is because the single quotes vs double quotes

No it's the generator expressions, but I'll explain more in the PR and probably create a follow-up issue.

Btw all contributing docs mention the `develop` branch instead of `develop2`, which I thought was weird.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/15922 will fix that, thanks for raising it too!

**Comment by memsharded** [comment]
The PR in https://github.com/conan-io/conan/pull/15924 will be released in 2.2.2 today, as it could be considered a regression. Thanks again for reporting.

**Comment by stevenwdv** [comment]
Maybe you can re-open this for now?

**Comment by memsharded** [comment]
Sure re-opening.

It would be great to have a PR with a test similar to ``test_cmake_toolchain_cxxflags_multi_config()``, as seen in  https://github.com/conan-io/conan/pull/15924/files that fails. Probably the best would be a new test if it can add just new checks, or a full new test if changing something else.

**Comment by stevenwdv** [comment]
@memsharded What I don't understand yet is why the current test is passing (see https://github.com/conan-io/conan/pull/15924#issuecomment-2019903707)

**Comment by DenizThatMenace** [comment]
This is still broken in Conan 2.5.

But the fix is quite simple, as I already commented in https://github.com/conan-io/conan/pull/15924#issuecomment-2217357841:


```diff
--- a/conan/tools/cmake/toolchain/blocks.py
+++ b/conan/tools/cmake/toolchain/blocks.py
@@ -724,7 +724,7 @@ class ExtraFlagsBlock(Block):
         {% if defines %}
         {% if config %}
         {% for define in defines %}
-        add_compile_definitions($<$<CONFIG:{{config}}>:"{{ define }}">)
+        add_compile_definitions("$<$<CONFIG:{{config}}>:{{ define }}>")
         {% endfor %}
         {% else %}
         add_compile_definitions({% for define in defines %} "{{ define }}"{% endfor %})
```

**Comment by memsharded** [comment]
To consider these changes, we would need first a failing test that reproduces the issues.

I am trying to reproduce, and added ``tools.cmake.cmaketoolchain:generator=Ninja Multi-Config`` in the existing test, but it is still green and passing.

Maybe you want to try to do a change or a new test like https://github.com/conan-io/conan/blob/dd28339d2964593471ead06257b89fec5e4ab24e/test/functional/toolchains/cmake/test_cmake_toolchain.py#L1831, and submit a PR with it that showcases the error? That would help to clarify the issues.

**Comment by DenizThatMenace** [comment]
I have created a fix and test for this issue in PR #16637.
